### DISCUSSION TITLE
Ideas

### DISCUSSION THREAD

**Discussion by mgrachev** [dotenv-linter/dotenv-linter#23] [Ideas]
Hi there!

If you have any ideas on how to improve this project or what other checks we should add - welcome 😉

**Comment by mstruebing** [comment]
A dockerimage would be nice I think

**Comment by mgrachev** [comment]
> A dockerimage would be nice I think

Could you give more information about that? Do you want to run this linter using a Docker image?

**Comment by mstruebing** [comment]
Yes, an image with an already compiled binary which I can just use without the need to install `dotenv-linter` or any dependencies (like rust for example).

**Comment by qelphybox** [comment]
Disabling/enabling comments. It's a good option, i think some users may want to disable some checkers 
e.g.
```env
# linter:disable UnorderedKeysChecker
 
MY=val
OWN=val
ORDER=val
```

**Comment by qelphybox** [comment]
I think autofix option is awesome, that's what i really love in linters
We have an `.env`
```env
 LOGGER_LEVEL=info
RAILS_ENV
DB-NAME=testing
DEbUG_hTTP=true
DB_NAME=development

```

run `dotenv-linter --fix`

- remove leading spaces
- replace incorrect delimeter to "_"
- set '=' to empty keys
- remove duplicated keys, sets value of the last key   
- order keys

We'll get:
```env
DB_NAME=development
DEBUG_HTTP=true
LOGGER_LEVEL=info
RAILS_ENV=
```

**Comment by mgrachev** [comment]
> Yes, an image with an already compiled binary which I can just use without the need to install `dotenv-linter` or any dependencies (like rust for example).

Great idea! I've just created an issue to do that https://github.com/mgrachev/dotenv-linter/issues/34.

**Comment by mgrachev** [comment]
> Disabling/enabling comments. It's a good option, i think some users may want to disable some checkers
> e.g.
> 
> ```
> # linter:disable UnorderedKeysChecker
>  
> MY=val
> OWN=val
> ORDER=val
> ```

I think it's a good idea! Could you open an issue and write your vision on how this will work?

**Comment by mgrachev** [comment]
> I think autofix option is awesome, that's what i really love in linters
> We have an `.env`
> 
> ```
>  LOGGER_LEVEL=info
> RAILS_ENV
> DB-NAME=testing
> DEbUG_hTTP=true
> DB_NAME=development
> ```
> 
> run `dotenv-linter --fix`
> 
> * remove leading spaces
> * replace incorrect delimeter to "_"
> * set '=' to empty keys
> * remove duplicated keys, sets value of the last key
> * order keys
> 
> We'll get:
> 
> ```
> DB_NAME=development
> DEBUG_HTTP=true
> LOGGER_LEVEL=info
> RAILS_ENV=
> ```

Awesome idea! Thank you! I will think about how to implement it.

**Comment by mstruebing** [comment]
Maybe it would be a good idea to give the user the ability to lint a specific file, i.e.: `dotenv-linter .my-env-file`

**Comment by artem-russkikh** [comment]
I think it would be good to add check for lines trailing spaces, what do you think?

Also it would be good to skip blank lines. Blank lines could be used in env files for separation some blocks of variables. ENV file like this:
```env
FOO=BAR

BAR=FOO
```
prints out `.env:2 The  key should be with a value or have an equal sign`

And I've created issue for filtering out comments https://github.com/mgrachev/dotenv-linter/issues/36

**Comment by mgrachev** [comment]
> Maybe it would be a good idea to give the user the ability to lint a specific file, i.e.: `dotenv-linter .my-env-file`

You're right! I've already thought about it. I will open an issue to do that a bit later.

**Comment by mgrachev** [comment]
> I think it would be good to add check for lines trailing spaces, what do you think?
> 
> Also it would be good to skip blank lines. Blank lines could be used in env files for separation some blocks of variables. ENV file like this:
> 
> ```
> FOO=BAR
> 
> BAR=FOO
> ```
> 
> prints out `.env:2 The key should be with a value or have an equal sign`
> 
> And I've created issue for filtering out comments #36

Good ideas! Thank you!

**Comment by artem-russkikh** [comment]
Change exit code to nonzero if some checks are failing – it will be useful for CI

**Comment by mgrachev** [comment]
> Maybe it would be a good idea to give the user the ability to lint a specific file, i.e.: `dotenv-linter .my-env-file`

Add ability to include files to check - https://github.com/mgrachev/dotenv-linter/issues/45
Add ability to exclude files from check - https://github.com/mgrachev/dotenv-linter/issues/46

**Comment by mgrachev** [comment]
> Change exit code to nonzero if some checks are failing – it will be useful for CI

Good catch! Thank you! https://github.com/mgrachev/dotenv-linter/issues/47

**Comment by Neodelf** [comment]
I don't know exactly about the pros of this feature, but what do you think about creation `.env.example` file which gathers all variables from all `.env.*` (i.e. `.env.test`) files?
I mean it will help keep all available vars in one place.
In addition, this feature can insert an example of the meaning of each variable by regex-ing values from files.

**Comment by mgrachev** [comment]
Thank you!
I think it will be useful, but the main task of dotenv-linter is to check `.env` files.
So, I think this feature should be implemented in a separate tool.

**Comment by GaboFDC** [comment]
what about allowing to have a config file, where the excluded checks are defned?

**Comment by DDtKey** [comment]
> what about allowing to have a config file, where the excluded checks are defned?

Thanks, I think that's a good idea!  
In general, we have already partially begun to move in this direction (#237)

**Comment by alshdavid** [comment]
It might be out of the scope of this project but it would be great if your utility could also inject the values into the current shell/bash/cmd/etc session.

It's really hard to import environment variables from a dotenv file to a OS in a way that is platform agnostic

Something like:

```
dotenv-linter --inject ./values.env
```

**Comment by mstruebing** [comment]
Hey @alshdavid, thank you for your idea but to be honest I think this is out of scope as you already mentioned.
And to my knowledge a single `source values.env` should make the environment variables available in the current shell session at least in unixy environments.

**Comment by bryantbiggs** [comment]
great tool btw, thank you! it would be cool if it had the ability to diff across files to compare say `.env.local` and `.env.dev` or even compare all `.env*` in a directory to know which files might be missing values

**Comment by mgrachev** [comment]
> great tool btw, thank you! it would be cool if it had the ability to diff across files to compare say `.env.local` and `.env.dev` or even compare all `.env*` in a directory to know which files might be missing values

Thank you! We will think about it soon.

**Comment by mstruebing** [comment]
> great tool btw, thank you! it would be cool if it had the ability to diff across files to compare say .env.local and .env.dev or even compare all .env* in a directory to know which files might be missing values

A cool idea, I really like that to be honest.

**Comment by ben-eb** [comment]
We use https://www.npmjs.com/package/dotenv-safe to make sure that any missing environment variables are flagged at runtime. It might be nice to have an option in this tool to check a `.env` file against an example env to make sure that there are no missing variables. 🙂

**Comment by mgrachev** [comment]
> We use https://www.npmjs.com/package/dotenv-safe to make sure that any missing environment variables are flagged at runtime. It might be nice to have an option in this tool to check a `.env` file against an example env to make sure that there are no missing variables. 🙂

Thank you! It seems to me that @mstruebing has already opened a PR with similar functionality https://github.com/dotenv-linter/dotenv-linter/pull/282 🤔

**Comment by ben-eb** [comment]
@mgrachev Thanks, that's great news! 👍

**Comment by mgrachev** [comment]
We are happy to announce the release of dotenv-linter - [v3.0.0](https://github.com/dotenv-linter/dotenv-linter/releases/tag/v3.0.0) 🥳
Thanks to everyone who contributed to this release 🙏

**Comment by eric-burel** [comment]
Hey folks, crossposting from "modenv" project: https://github.com/kurtbuilds/modenv/issues/4

Do you think `dotenv-linter` could have a wider scope than just linting? For instance `modenv` let's you add variables to file easily via CLI, and I've proposed in the linked issue to extend it to a .env backup system. I am kinda looking for a non-cloud alternative to [Doppler](https://www.doppler.com/).

However, I prefer having one go-to solution than multiple projects, so I could instead try to add this feature to `dotenv-linter` which sounds more popular. Would you be interested in a PR implementing this backup feature?

This comment is also related, .env injection is a key feature of Doppler: https://github.com/dotenv-linter/dotenv-linter/discussions/23#discussioncomment-164389

**Comment by jblotus** [comment]
I found out by accident that passing a directory as an the first argument to compare you can compare all .env files

Windows:
```
dotenv-linter compare %cd% .env.example
```
Linux
```
dotenv-linter compare . .env.example
```

Output:
```
Comparing .env.dev
Comparing .env.qa
Comparing .env.local
Comparing .env.prod
```

Hope that helps someone
